Incorrect span label when non-() else-less `if` is tail expression in `-> ()` function

### Code

```Rust
fn main() {
    if true {
        ""
    }
}
```


### Current output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:3:9
  |
1 | fn main() {
  |          - expected `()` because of default return type
2 |     if true {
3 |         ""
  |         ^^ expected `()`, found `&str`
```


### Desired output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:3:9
  |
2 | /     if true {
3 | |         ""
  | |         ^^ expected `()`, found `&str`
4 | |     }
  | |_____- expected this to be `()`
```


### Rationale and extra context

When the `if` is *not* the tail expression, we provide the right output. We might just have to change the order of operation when looking for additional context to print.

### Other cases

`if true { "" } else { "" }` as tail expression has the same problem.

### Rust Version

All checked, current version 1.78.

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"cardigan1008"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

@cardigan1008 I believe you can just modify `get_return_block` to additionally check whether the node is an else-less `if` (adding an arm to the match like `Node::Block(Block { expr: Some(e), .. }) if matches!(e.kind, ExprKind::If(_, _, None)) => return None,`) with no adverse effects, but you'll have to try it out to see if there's any covered use of `get_return_block` that *expects* that case to continue climbing the hir tree. Just run `./x.py test tests/ui --bless` and that will modify the `.stderr` files to see the effects of your change.

The other case:
```
fn main() {
    if true {
        ""
    } else {
        ""
    }
}
```
```
error[E0308]: mismatched types
 --> src/main.rs:3:9
  |
1 | fn main() {
  |          - expected `()` because of default return type
2 |     if true {
3 |         ""
  |         ^^ expected `()`, found `&str`

error[E0308]: mismatched types
 --> src/main.rs:5:9
  |
1 | fn main() {
  |          - expected `()` because of default return type
...
5 |         ""
  |         ^^ expected `()`, found `&str`
```

I think is ok. It could do with a more targeted error that will be more involved (to emit a single error instead of multiple), but I wouldn't consider it in scope for this ticket.

@estebank Thanks for you kind and clear instructions! I've ran `./x.py test tests/ui --bless` and it turned out that this change has no effect on the other cases. Considering the other case you mentioned, in my opinion, I think it may be better to emit multiple errors. Here is a case:
```rust
fn main() {
    if true {
        ""
    } else if false {
        
    } else {
        ""
    }
}
```
```
error[E0308]: mismatched types
 --> src/main.rs:3:9
  |
1 | fn main() {
  |          - expected `()` because of default return type
2 |     if true {
3 |         ""
  |         ^^ expected `()`, found `&str`

error[E0308]: mismatched types
 --> src/main.rs:5:9
  |
1 | fn main() {
  |          - expected `()` because of default return type
...
5 |         ""
  |         ^^ expected `()`, found `&str`

error[E0308]: mismatched types
 --> src/main.rs:9:9
  |
1 | fn main() {
  |          - expected `()` because of default return type
...
9 |         ""
  |         ^^ expected `()`, found `&str`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (bin "playground") due to 3 previous errors
```
In this case, would aggregating errors into a single message make the correction less clear? I think the multiple errors might be a good reminder for users to fix errors in several specific branches. Maybe I misunderstand your meaning about the single error. Sorry for that.

--- a/compiler/rustc_middle/src/hir/map/mod.rs
+++ b/compiler/rustc_middle/src/hir/map/mod.rs
@@ -549,6 +549,7 @@ impl<'hir> Map<'hir> {
                     Node::Block(Block { expr: None, .. }) => return None,
                     // The current node is not the tail expression of its parent.
                     Node::Block(Block { expr: Some(e), .. }) if hir_id != e.hir_id => return None,
+                    Node::Block(Block { expr: Some(e), ..}) if matches!(e.kind, ExprKind::If(_, _, None)) => return None,
                     _ => {}
                 }
             }
@@ -563,9 +564,6 @@ impl<'hir> Map<'hir> {
                     // We verify that indirectly by checking that the previous node is the
                     // current node's body
                     if node.body_id().map(|b| b.hir_id) == prev_hir_id =>  {
-                        if let Node::Expr(Expr { kind: ExprKind::Block(_, _), ..}) = self.tcx.hir_node(prev_hir_id.unwrap()) {
-                            return None;
-                        }
                         return Some(hir_id)
                 }
                 // Ignore `return`s on the first iteration

--- a/compiler/rustc_middle/src/hir/map/mod.rs
+++ b/compiler/rustc_middle/src/hir/map/mod.rs
@@ -563,6 +563,9 @@ impl<'hir> Map<'hir> {
                     // We verify that indirectly by checking that the previous node is the
                     // current node's body
                     if node.body_id().map(|b| b.hir_id) == prev_hir_id =>  {
+                        if let Node::Expr(Expr { kind: ExprKind::Block(_, _), ..}) = self.tcx.hir_node(prev_hir_id.unwrap()) {
+                            return None;
+                        }
                         return Some(hir_id)
                 }
                 // Ignore `return`s on the first iteration

--- a/tests/ui/return/tail-expr-if-as-return.rs
+++ b/tests/ui/return/tail-expr-if-as-return.rs
@@ -1,0 +1,5 @@
+fn main() {
+    if true {
+        "" //~ ERROR mismatched types [E0308]
+    }
+}

--- a/tests/ui/return/tail-expr-if-as-return.stderr
+++ b/tests/ui/return/tail-expr-if-as-return.stderr
@@ -1,0 +1,12 @@
+error[E0308]: mismatched types
+  --> $DIR/tail-expr-if-as-return.rs:3:9
+   |
+LL | /     if true {
+LL | |         ""
+   | |         ^^ expected `()`, found `&str`
+LL | |     }
+   | |_____- expected this to be `()`
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0308`.

